README.md
# indicator-china-weather

The weather data are from the heweather API s6 version.

![](./doc/weather_zh_CN.png)

### Author's Home Page
 
[Eight Plus &rarr;](https://eightplus.github.io/)



### v1.0 test url

+ http://service.ubuntukylin.com:8001/weather/api/1.0/observe/101250101/

+ http://service.ubuntukylin.com:8001/weather/api/1.0/heweather_forecast/101250101/

+ http://service.ubuntukylin.com:8001/weather/pingnetwork/

+ http://service.ubuntukylin.com:8001/weather/pinginformation/


### v2.0 test url

+ http://service.ubuntukylin.com:8001/weather/api/2.0/heweather_observe_s6/101250101

+ http://service.ubuntukylin.com:8001/weather/api/2.0/heweather_forecast_s6/101250101


### Internationalization

1. lupdate indicator-china-weather.pro
2. linguist translation/indicator-china-weather_zh_CN.ts
3. lrelease indicator-china-weather.pro


### Lintian

lintian -i -EvIL +pedantic --verbose ../indicator-china-weather_3.0.0_amd64.changes


debian/changelog
indicator-china-weather (3.1.2-0ubuntu1) impish; urgency=medium

  * New upstream release.

 -- handsome_feng <jianfengli@ubuntukylin.com>  Thu, 05 Aug 2021 16:37:26 +0800

indicator-china-weather (3.0.4.1-0ubuntu1) hirsute; urgency=medium

  * Add QPainterPath to avoid FTBFS.

 -- handsome_feng <jianfengli@ubuntukylin.com> Fri, 09 Apr 2021 20:16:18 +0800

indicator-china-weather (3.0.4-0ubuntu1) focal; urgency=medium

  * New bugfix release:
    - Fix the 404 error by set the default city and fix the wrong url.
      (LP: #1870951)
    - Don't start if locale is no chinese since it is not friendly to
      non-chinese users now.
  * debian:
    - Bump standards-versition to 4.5.0.
    - Use debhelper-compat instead of debhelper.
    - Set Rules-Requires-Root to no.

 -- handsome_feng <jianfengli@ubuntukylin.com>  Fri, 10 Apr 2020 16:23:19 +0800

indicator-china-weather (3.0.3-0ubuntu2) focal; urgency=medium

  * No-change rebuild for libgcc-s1 package name change.

 -- Matthias Klose <doko@ubuntu.com>  Wed, 25 Mar 2020 19:06:56 +0100

indicator-china-weather (3.0.3-0ubuntu1) eoan; urgency=medium

  [ Li Xiang ]
  * Modify the AboutDialog.
  * Update translations for MainWindow.
  * Access icon from theme.
  * Change `PKGCONFIG += geoip` to `LIBS += -lGeoIP` for 19.10.
  * Support for Auto-Location of Cities.
  * Add the masking layer when acquiring data.
  * Fixed the bug that display location error.
  * Update translations.
  * Handling mouse events on pallets.
  * PrimaryScreen display.
  * Hide window when focusOutEvent occurred.
  * Mainwindow moves with the position of panel.
  * Fixed weekly error.
  * Used the latest version of heweather API.
  * New UI with Qt5.
  * Fixed the bug that Refresh time error.
  * Show on the bottom right of the screen when on UKUI.
  * Modified the CSS style of cityitemwidget.
  * Modified the path of translated pm files and the timing of generation.

  [ handsome_feng ]
  * debian/compat: 11 -> 12
  * debian/control:
    - Bump standards-vertion to 4.4.0.
    - Bump debhelper requirement for compat

 -- handsome_feng <jianfengli@ubuntukylin.com>  Mon, 12 Aug 2019 15:24:19 +0800

indicator-china-weather (2.2.8-0ubuntu1) bionic; urgency=medium

  * Remove "Recommends: python-apport" in the control file.
    (LP: #1763732)
  * Fix the bug that the aboutdialog failed to popped out.

 -- handsome_feng <jianfengli@ubuntukylin.com>  Tue, 17 Apr 2018 10:29:19 +0800

indicator-china-weather (2.2.7-0ubuntu1) bionic; urgency=medium

  * Remove "StartupNotify=true" in the desktop file.
  * debian/compat: 9 -> 11
  * debian/control:
    - Bump debhelper requirement for compat
    - Bump std-ver: 3.9.5 -> 4.1.3
  * Add debian/watch file.

 -- handsome_feng <jianfengli@ubuntukylin.com>  Mon, 19 Mar 2018 14:04:57 +0800

indicator-china-weather (2.2.6-0ubuntu1) artful; urgency=medium

  * Optimize data processing at weather server.
  * Optimize city additions with GObject.idle.

 -- lixiang <lixiang@kylinos.cn>  Thu, 31 Aug 2017 17:34:05 +0800

indicator-china-weather (2.2.5-0ubuntu1) artful; urgency=medium

  * Determine the whether data is empty or not.

 -- lixiang <lixiang@kylinos.cn>  Tue, 29 Aug 2017 17:06:02 +0800

indicator-china-weather (2.2.4-0ubuntu1) artful; urgency=medium

  * Modify forecast with thread daemon and idle.
  * Add retry button when access forecast data failed.

 -- lixiang <lixiang@kylinos.cn>  Wed, 09 Aug 2017 22:14:26 +0800

indicator-china-weather (2.2.3-0ubuntu1) artful; urgency=medium

  * Change weather api.
  * More detailed three day weather forecast.
  * Weather forecast with thread.

 -- lixiang <lixiang@kylinos.cn>  Wed, 28 Jun 2017 13:52:16 +0800

indicator-china-weather (2.2.1-0ubuntu1) zesty; urgency=medium

  * [ Jeremy Bicha <jbicha@ubuntu.com> ] debian/control:
    - Depend on gir1.2-appindicator3-0.1 and python-gi instead of no longer
      used python-appindicator, python-gobject, and python-gtk2
  * Resolved lintian warning.

 -- lixiang <lixiang@kylinos.cn>  Mon, 20 Mar 2017 11:37:03 +0800

indicator-china-weather (2.1.5-0ubuntu1) xenial; urgency=low

  * Modified control file for python 2.7.

 -- Kobe Lee (kylinkobe) <xiangli@ubuntukylin.com>  Mon, 14 Mar 2016 10:06:40 +0800

indicator-china-weather (2.1.4-0ubuntu1) xenial; urgency=low

  * Rewrite UI with css.

 -- Kobe Lee (kylinkobe) <xiangli@ubuntukylin.com>  Fri, 22 Jan 2016 09:24:04 +0800

indicator-china-weather (2.1.3-0ubuntu1) wily; urgency=low

  * Repair the real-time temperature and humidity.
  * Optimize the forecast interface.
  * Modified Database.

 -- Kobe Lee (kylinkobe) <xiangli@ubuntukylin.com>  Thu, 23 Jul 2015 11:25:46 +0800

indicator-china-weather (2.1.2-0ubuntu1) utopic; urgency=low

  * Replace gtk with Gtk.
  * Modify desktop.

 -- Kobe Lee (kylinkobe) <xiangli@ubuntukylin.com>  Tue, 30 Dec 2014 10:05:00 +0800

indicator-china-weather (2.1.1-0ubuntu1) utopic; urgency=low

  * Adjust the time range of automatically update, and set the default rate 15(LP: #1350675).
  * Solved the crashes problem without network (LP: #1334118).
  * Add notification.

 -- Kobe Lee (kylinkobe) <xiangli@ubuntukylin.com>  Wed, 20 Aug 2014 08:59:25 +0800


indicator-china-weather (2.1.0-0ubuntu1) trusty; urgency=low

  * Show temperature (LP: #1325791).
  * Make the Indicator icon clear.

 -- Kobe Lee (kylinkobe) <xiangli@ubuntukylin.com>  Mon, 16 Jun 2014 16:48:36 +0800

indicator-china-weather (2.0.0-0ubuntu1) trusty; urgency=low

  * Use pyqt to show widget and access SmartWeather API to get weather data.
  * Make indicator icon normally after auto updated (LP: #1323565).
  * Resume enable status of delete city button (LP: #1324020).
  * Make it normal when add and delete city list.
  * Deal with status when weather data is empty.

 -- Kobe Lee (kylinkobe) <xiangli@ubuntukylin.com>  Thu, 29 May 2014 17:43:12 +0800

indicator-china-weather (1.1.0-0ubuntu1) trusty; urgency=low

  * New upstream release (LP: #1282914)
    - Avoid adding the same city.

src/data.h
/*
 * Copyright (C) 2020, KylinSoft Co., Ltd.
 *
 * Authors:
 *  Kobe Lee    lixiang@kylinos.cn/kobe24_lixiang@126.com
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; version 3.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

#ifndef DATA_H
#define DATA_H

#include <QObject>
#include <QMetaType>

struct ItemData
{
    QString cityId;
    QString cityName;
    QString cityProvince;
};

struct City
{
    QString id;
    QString name;
};


struct AqiAir
{
    QString id;
    QString aqi;
    QString qlty;
    QString main;
    QString pm25;
    QString pm10;
    QString no2;
    QString so2;
    QString co;
    QString o3;
};

struct ObserveWeather
{
    QString id;
    QString city;
    QString updatetime;
    QString air;//空气质量
    QString cloud;//云量
    QString cond_code;//实况天气状况代码 	100
    QString cond_txt;//实况天气状况描述 晴
    QString fl;//体感温度，默认单位：摄氏度 	23
    QString hum;//相对湿度 	40
    QString pcpn;//降水量(毫米 mm) 	0
    QString pres;//大气压强(百帕 hPa) 	1020
    QString tmp;//温度，默认单位：摄氏度℃ 	21
    QString vis;//能见度，默认单位：公里 km 	10
    QString wind_deg;//风向360角度 	305
    QString wind_dir;//风向 	西北
    QString wind_sc;//风力 	3 (0:静风  1:1级风)
    QString wind_spd;//风速，公里/小时 km/h 	15
    QString serveTime;//天气信息的更新时间
};

struct ForecastWeather
{
    QString cond_code_d;//白天天气状况代码 	100
    QString cond_code_n;//晚间天气状况代码 	100
    QString cond_txt_d;//白天天气状况描述 	晴
    QString cond_txt_n;//晚间天气状况描述 	晴
    QString forcast_date;//预报日期 	2013-12-30
    QString hum;//相对湿度 	37
    //QString mr_ms;//月升时间 	04:47   月落时间 	14:59
    QString mr;//月升时间 	04:47
    QString ms;//月落时间 	14:59
    QString pcpn;//降水量 	0
    QString pop;//降水概率 	0
    QString pres;//大气压强 	1018
    //QString sr_ss;//日出时间 	07:36   日落时间 	16:58
    QString sr;//日出时间 	07:36   日落时间 	16:58
    QString ss;//日出时间 	07:36   日落时间 	16:58
    QString tmp_max;//最高温度 	4
    QString tmp_min;//最低温度 	-5
    QString uv_index;//紫外线强度指数 	3
    QString vis;//能见度，单位：公里 	10
    QString wind_deg;//风向360角度 	310
    QString wind_dir;//风向 	西北风
    QString wind_sc;//风力 	1-2 (级风)
    QString wind_spd;//风速，公里/小时 	14
    QString date;//日期
    QString dateTime;//当前城市信息更新日期
};

struct LifeStyle
{
    QString air_brf;
    QString air_txt;
    QString comf_brf;
    QString comf_txt;
    QString cw_brf;
    QString cw_txt;
    QString drsg_brf;
    QString drsg_txt;
    QString flu_brf;
    QString flu_txt;
    QString sport_brf;
    QString sport_txt;
    QString trav_brf;
    QString trav_txt;
    QString uv_brf;
    QString uv_txt;
};

struct LocationData {
    QString id;
    QString city_en;
    QString city;
    QString country_en;
    QString country;
    QString province_en;
    QString province;
    QString admin_district_en;
    QString admin_district;
    QString shorthand;
};

class CitySettingData
{
public:
    bool active;

    QString id;
    QString name;
    QString temperature;
    QString icon;

public:
    void setActive(bool b) { active = b; }
};

inline bool operator ==(const City &data, const City &other) {
    return data.id == other.id;
}

inline bool operator ==(const CitySettingData &data, const CitySettingData &other) {
    return data.id == other.id;
}

inline bool operator ==(const LocationData &data, const LocationData &that)
{
    return data.id == that.id || (data.admin_district == that.admin_district && data.city == that.city);
}

Q_DECLARE_METATYPE(ItemData)
Q_DECLARE_METATYPE(City)
Q_DECLARE_METATYPE(ObserveWeather)
Q_DECLARE_METATYPE(ForecastWeather)
Q_DECLARE_METATYPE(LifeStyle)
Q_DECLARE_METATYPE(LocationData)
Q_DECLARE_METATYPE(CitySettingData)

#endif // DATA_H
